File: src/main/java/org/processmining/utils/helpers/PetriNetGenerationHelper.java
package org.processmining.utils.helpers;

import org.processmining.models.*;
import org.processmining.models.abstract_net_representation.Place;
import org.processmining.models.abstract_net_representation.Token;
import org.processmining.models.abstract_net_representation.Transition;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * Created by Иван on 27.10.2014.
 */
public abstract class PetriNetGenerationHelper<T extends Place, K extends Transition, F extends Token> extends BaseGenerationHelper<T, K, F>
{

    public PetriNetGenerationHelper(Collection<T> initialMarking, Collection<T> finalMarking, Collection<K> allTransitions, Collection<T> allPlaces, GenerationDescription description)
    {
        super(initialMarking, finalMarking, allTransitions, allPlaces, description);
    }

    @Override
    public Movable chooseNextMovable()
    {
        List<Transition> enabledTransitions = findEnabledTransitions();

        Movable movable;

        if (enabledTransitions.isEmpty() && getExtraMovables().isEmpty())
        {
            return null;
        }
        else
        {
            if (enabledTransitions.isEmpty())
            {
                movable = pickRandomMovable(getExtraMovables());
            }
            else
            {
                if (getExtraMovables().isEmpty())
                {
                    movable = pickRandomMovable(enabledTransitions);
                }
                else
                {
                    boolean moveThroughTransition = random.nextBoolean();
                    if (moveThroughTransition)
                    {
                        movable = pickRandomMovable(enabledTransitions);
                    }
                    else
                    {
                        movable = pickRandomMovable(getExtraMovables());
                    }
                }
            }
        }

        return movable;
    }

    protected List<Transition> findEnabledTransitions()
    {
        List<Transition> enabledTransitions = new ArrayList<Transition>();

        for (Transition transition : getAllModelMovables())
        {
            if (transition.checkAvailability())
            {
                enabledTransitions.add(transition);
            }
        }
        return enabledTransitions;
    }

}

File: src/main/java/org/processmining/utils/helpers/PetriNetGenerationHelper.kt
package org.processmining.utils.helpers

import org.processmining.models.*
import org.processmining.models.abstract_net_representation.Place
import org.processmining.models.abstract_net_representation.Token
import org.processmining.models.abstract_net_representation.Transition

import java.util.ArrayList

/**
 * Created by Иван on 27.10.2014.
 */

abstract class PetriNetGenerationHelper<T : Place<*>, K : Transition<*>, F : Token>(
        initialMarking: Collection<T>,
        finalMarking: Collection<T>,
        allTransitions: Collection<K>,
        allPlaces: Collection<T>,
        description: GenerationDescription
) : BaseGenerationHelper<T, K, F>(initialMarking, finalMarking, allTransitions, allPlaces, description) {
    
    override fun chooseNextMovable(): Movable? {
        val enabledTransitions = findEnabledTransitions()
        
        val movable: Movable
        
        if (enabledTransitions.isEmpty() && extraMovables.isEmpty()) {
            return null
        } else {
            if (enabledTransitions.isEmpty()) {
                movable = pickRandomMovable(extraMovables)
            } else {
                if (extraMovables.isEmpty()) {
                    movable = pickRandomMovable(enabledTransitions)
                } else {
                    val moveThroughTransition = random.nextBoolean()
                    if (moveThroughTransition) {
                        movable = pickRandomMovable(enabledTransitions)
                    } else {
                        movable = pickRandomMovable(extraMovables)
                    }
                }
            }
        }
        
        return movable
    }
    
    protected fun findEnabledTransitions() = allModelMovables.filter { it.checkAvailability() }
    
}
